### PR TITLE
Update GCS hook to get crc32c hash for CMEK-protected objects

### LINKED ISSUE #34980: GCSSynchronizeBucketsOperator does not retrieve crc32c hash for CMEK objects
### Apache Airflow version

2.7.2

### What happened

The GCSSynchronizeBucketsOperator eventually calls the `_prepare_sync_plan` function in GCS' `hooks.py`. This function retrieves objects in the buckets using the [`list_blobs` method](https://github.com/apache/airflow/blob/8fdf3582c2967161dd794f7efb53691d092f0ce6/airflow/providers/google/cloud/hooks/gcs.py#L1307). However, at present, the Cloud Storage Objects.List API does not return the crc32c for CMEK-protected objects. As per the [GCP public docs](https://cloud.google.com/storage/docs/encryption/customer-managed-keys#restrictions), "The CRC32C checksum and MD5 hash of objects encrypted with customer-managed encryption keys are not returned when listing objects with the JSON API." 

As a result, if an object is CMEK-protected, its crc32c value is always None, leading to incorrect synchronization ([crc32c comparison](https://github.com/apache/airflow/blob/8fdf3582c2967161dd794f7efb53691d092f0ce6/airflow/providers/google/cloud/hooks/gcs.py#L1330)).

### What you think should happen instead

This should be handled by making an Objects.Get call to retrieve the crc32c for CMEK'd objects.

### How to reproduce

1. Create a GCP Cloud Key Management Service (KMS) key.
2. Create two Cloud Storage buckets with a default bucket CMEK key.
3. Upload an object with the same name but different contents to each bucket.
4. Run the GCSSynchronizeBucketsOperator with one bucket as source and one as destination, and `allow_overwrite=True`. Since a file is found with the same name in each, the crc32c will be compared. Since they are both None, they are seen as equal, and the source object does not overwrite the destination one.

### Operating System

Linux / Cloud Composer

### Versions of Apache Airflow Providers

_No response_

### Deployment

Google Cloud Composer

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by dmedora**
Closes: #34980 

Updated _prepare_sync_plan function to get blobs using the Objects Get API (.get_blob(..)) so that their crc32c value is included in the returned Blob object. As discussed in #34980, the current Objects List method does not return this for CMEK-protected objects, leading to inaccurate hash comparisons.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by heaven00** [comment]
Shouldn't this need a test against whatever that blob should be?

**Comment by dmedora** [comment]
> Shouldn't this need a test against whatever that blob should be?

Not sure I understand, could you please explain further or with an example? Thanks.

**Comment by heaven00** [comment]
sure, 

so the function that has been changed is here is a private function, so we can't test the logic that has been added but from what I can understand there should be some behavioral changes to the `sync` function https://github.com/apache/airflow/pull/34982/files#diff-82854006b5553665046db26d43a9dfa90bec78d4ba93e2d2ca7ff5bf632fa624R1179


in terms of the data it is parsing from the blob and in my opinion there should be respective tests added to https://github.com/RNHTTR/airflow/blob/78d1031b34d70784d22fc1360c6f3deeb2c66a2c/airflow/providers/google/cloud/hooks/gcs.py so that the behavior is captured.

File: airflow/providers/google/cloud/hooks/gcs.py [hunk 1]
```diff
@@ -1326,6 +1326,11 @@ def _prepare_sync_plan(
         for current_name in names_to_check:
             source_blob = source_names_index[current_name]
             destination_blob = destination_names_index[current_name]
+            # If either object is CMEK-protected, use the Cloud Storage Objects Get API to retrieve them so that the crc32c is included
+            if source_blob.kms_key_name != None:
```
Comment (Taragolis) [inline airflow/providers/google/cloud/hooks/gcs.py]:
```suggestion
            if source_blob.kms_key_name:
```

File: airflow/providers/google/cloud/hooks/gcs.py [hunk 2]
```diff
@@ -1326,6 +1326,11 @@ def _prepare_sync_plan(
...
+                source_blob = source_bucket.get_blob(source_blob.name)
+            if destination_blob.kms_key_name != None:
```
Comment (Taragolis) [inline airflow/providers/google/cloud/hooks/gcs.py]:
```suggestion
            if destination_blob.kms_key_name:
```

File: airflow/providers/google/cloud/hooks/gcs.py (hunk 1 above)
Comment (dmedora) [reply airflow/providers/google/cloud/hooks/gcs.py]: Updated in latest commit, thanks.

File: airflow/providers/google/cloud/hooks/gcs.py (hunk 2 above)
Comment (dmedora) [reply airflow/providers/google/cloud/hooks/gcs.py]: Updated in latest commit, thanks.

**Comment by Taragolis** [comment]
Need to fix [static checks](https://github.com/apache/airflow/blob/main/STATIC_CODE_CHECKS.rst#id5) and failed test

**Comment by dmedora** [comment]
Thanks, still figuring out how to write the test correctly.

**Comment by Taragolis** [comment]
Are you worry about that you can't test internal static method `_prepare_sync_plan`?
I don't see any problem here because it is internal part of Google Provider so you could also test it as part of Google Provider

File: tests/providers/google/cloud/hooks/test_gcs.py [hunk 3]
```diff
@@ -1253,6 +1253,45 @@ def test_should_overwrite_files(self, mock_get_conn, mock_delete, mock_rewrite,
             any_order=True,
         )
         mock_copy.assert_not_called()
+
```
Comment (eladkal) [inline tests/providers/google/cloud/hooks/test_gcs.py]:
```suggestion

```

File: tests/providers/google/cloud/hooks/test_gcs.py [hunk 4]
```diff
@@ -1285,6 +1285,44 @@ def test_should_overwrite_files(self, mock_get_conn, mock_delete, mock_rewrite,
             any_order=True,
         )
         mock_copy.assert_not_called()
+    @mock.patch(GCS_STRING.format("GCSHook.copy"))
```
Comment (eladkal) [inline tests/providers/google/cloud/hooks/test_gcs.py]:
```suggestion
        mock_copy.assert_not_called()

    @mock.patch(GCS_STRING.format("GCSHook.copy"))
```

**Review by shahar1 — COMMENTED** [review]
The test `TestSyncGcsHook.test_should_overwrite_cmek_files` should be fixed.
Besides - LGTM.

**Comment by eladkal** [comment]
@dmedora can you please fix the test?

**Pull request closed without merging**